source/common/upstream/ring_hash_lb.cc

```diff
@@ -58,36 +64,118 @@ HostConstSharedPtr RingHashLoadBalancer::Ring::chooseHost(uint64_t h) const {
   }
 }
 
+// TODO(mergeconflict): Determine whether the locality weights we get here are already adjusted
+//                      for partial availability (as in HostSetImpl::effectiveLocalityWeight), and
+//                      promote this into ThreadAwareLoadBalancerBase or HashingLoadBalancer so
+//                      Maglev LB can use it (see #5982).
+namespace {
+
+void normalizeHostWeights(const HostVector& hosts, double normalized_locality_weight,
+                          std::unordered_map<HostConstSharedPtr, double>& normalized_weights,
+                          double& min_normalized_weight) {
+  uint32_t sum = 0;
+  for (const auto& host : hosts) {
+    sum += host->weight();
+  }
+
+  for (const auto& host : hosts) {
+    const double weight = host->weight() * normalized_locality_weight / sum;
+    normalized_weights[host] = weight;
```

Comment:
The same host will appear in multiple priorities in the future as a result of some work that @snowp is doing. Could the same host also appear in multiple localities?

Reply:
That's an interesting question... I don't know if it's actually possible, but I can certainly code this as if it were.

Reply:
I think that would be desirable, thanks.

Reply:
Alright, @htuch and I decided to punt this issue for now, since it won't break any existing users.